### ISSUE TITLE
Sort out the test suite

### ISSUE THREAD

**Issue by simoncozens** [harfbuzz/harfbuzz#2754]
The Harfbuzz test suite is *amazing*, and could potentially form the basis of the kind of shaping spec [discussed elsewhere](https://github.com/w3c/font-text-cg/issues/11). However, looking through it I've noticed areas that need improvement. I'm happy to look into these:

- [ ] Find out why commented-out aots tests are commented out, and fix them.
- [ ] Ensure we have adequate tests for all basic functionality (GSUB1-8 all formats, GPOS1-9 all formats).
- [ ] Specific examples: the reverse-sub test doesn't test that the substitutions are done in reverse order.
- [ ] Ensure that each script-specific shaper has adequate tests. (e.g. there's currently no explicit Hebrew or Thai tests.)

**Comment by behdad** [comment]
> * [ ] Find out why commented-out aots tests are commented out, and fix them.

> * [ ]  Ensure we have adequate tests for all basic functionality (GSUB1-8 all formats, GPOS1-9 all formats).

Is *really* hard to test all aspects of those. When you get to marks, mark classes, mark attachments, ligation of marks, etc. Is completely unspecified, and what HB does is slightly different from Uniscribe in certain cases (by design).

> * [ ]  Ensure that each script-specific shaper has adequate tests. (e.g. there's currently no explicit Hebrew or Thai tests.)

Right.  In `test/shaping/texts` there's a bunch of interesting text. The first plan was to have interesting text, and fonts, and test each font against a set of text that is applicable to it.

Then there was the mega test suite, shaping wordlist from Wikipedia per language (http://github.com/harfbuzz/harfbuzz-testing-wikipedia) against default font for such language on Windows, comparing to Uniscribe output, as described in: https://goo.gl/9eWCLy

But storing the expected output for that was so huge we never figured out how to do. @jfkthame was running that on Windows bots on the cloud, comparing the HB and Uniscribe outputs on the fly, and he built a portal for it.. but I don't think he runs that anymore.

**Comment by behdad** [comment]
> * [ ] Find out why commented-out aots tests are commented out, and fix them.

AOTS had made assumptions about unspecified parts of the spec. HarfBuzz sometimes makes different assumptions.

One particular difference also is that AOTS assumes shaping populates both `x_advance` and `y_advance`. In HarfBuzz we only populate one of those, depending on selected text direction.

**Comment by khaledhosny** [comment]
Are you working or still intending to work on this @simoncozens? I might have few cycles to spare and I’d like to work on improving test coverage (seeing the plethora of new OpenType implementations, each buggy in its own way, I hope improving the test coverage in HarfBuzz might encourage testing these new implementations against HarfBuzz test suite and hopefully improve them).

**Comment by behdad** [comment]
Same feels. I like better test suite as well. But if there's nothing actionable, there's no point keeping this open.

**Comment by simoncozens** [comment]
Realistically I don't think I'm going to do it any time soon; I will be coming back to the fontFeatures shaper eventually, but it's not a high priority right now. So please go ahead.

**Comment by khaledhosny** [comment]
> Specific examples: the reverse-sub test doesn't test that the substitutions are done in reverse order.

Yes, that test is testing #2467 specifically, more tests are needed but this test was not design to test the general reverse substitution handling.
